File: database/database-annotation-processor/src/test/java/ru/tinkoff/kora/database/common/annotation/processor/DbTestUtils.java
package ru.tinkoff.kora.database.common.annotation.processor;

import ru.tinkoff.kora.annotation.processor.common.TestUtils;
import ru.tinkoff.kora.database.annotation.processor.RepositoryAnnotationProcessor;

import java.lang.reflect.Constructor;

public class DbTestUtils {

    public static <T> T compile(Class<T> repository, Object... params) {
        try {
            var cl = TestUtils.annotationProcess(repository, new RepositoryAnnotationProcessor());
            @SuppressWarnings("unchecked")
            var clazz = (Class<? extends T>) cl.loadClass(repository.getPackageName() + ".$" + repository.getSimpleName() + "_Impl");
            assert clazz.getConstructors().length == 1;
            var constructor = (Constructor<? extends T>) clazz.getConstructors()[0];
            return constructor.newInstance(params);
        } catch (RuntimeException throwable) {
            throw throwable;
        } catch (Throwable throwable) {
            throw new RuntimeException(throwable);
        }
    }
    public static <T> Class<? extends T> compileClass(Class<T> repository) {
        try {
            var cl = TestUtils.annotationProcess(repository, new RepositoryAnnotationProcessor());
            @SuppressWarnings("unchecked")
            var clazz = (Class<? extends T>) cl.loadClass(repository.getPackageName() + ".$" + repository.getSimpleName() + "_Impl");
            return clazz;
        } catch (RuntimeException throwable) {
            throw throwable;
        } catch (Throwable throwable) {
            throw new RuntimeException(throwable);
        }
    }

}

File: database/database-symbol-processor/src/test/kotlin/ru/tinkoff/kora/database/symbol/processor/DbTestUtils.kt
@file:OptIn(KspExperimental::class)

package ru.tinkoff.kora.database.symbol.processor

import com.google.devtools.ksp.KspExperimental
import com.squareup.kotlinpoet.asClassName
import ru.tinkoff.kora.ksp.common.CompilationErrorException
import ru.tinkoff.kora.ksp.common.symbolProcess
import java.lang.reflect.Constructor
import kotlin.reflect.KClass

object DbTestUtils {
    fun <T : Any> compile(repository: KClass<T>, vararg params: Any?): T {
        return try {
            val cl = symbolProcess(repository, RepositorySymbolProcessorProvider())
            val clazz = cl.loadClass(repository.asClassName().packageName + ".$" + repository.simpleName + "_Impl")
            assert(clazz.constructors.size == 1)
            val constructor = clazz.constructors[0] as Constructor<out T>
            constructor.newInstance(*params)
        } catch (e: CompilationErrorException) {
            throw e
        } catch (throwable: RuntimeException) {
            throw throwable
        } catch (throwable: Throwable) {
            throw RuntimeException(throwable)
        }
    }
    fun <T: Any> compileClass(repository: KClass<T>): KClass<out T> {
        try {
            val cl = symbolProcess(repository, RepositorySymbolProcessorProvider())
            return cl.loadClass(repository.asClassName().packageName + ".$" + repository.simpleName + "_Impl").kotlin as KClass<out T>
        } catch (e: CompilationErrorException) {
            throw e
        } catch (throwable: RuntimeException) {
            throw throwable
        } catch (throwable: Throwable) {
            throw RuntimeException(throwable)
        }
    }
}
